Compiler error? `##[error]HttpRequestStreamReaderReadLineBenchmark.cs(6,34): error CS0234: (NETCORE_ENGINEERING_TELEMETRY=Build) The type or namespace name 'Jobs' does not exist in the namespace 'BenchmarkDotNet.Attributes' (are you missing an assembly reference?)
`

--- a/src/Http/WebUtilities/src/HttpRequestStreamReader.cs
+++ b/src/Http/WebUtilities/src/HttpRequestStreamReader.cs
@@ -155,50 +155,8 @@ namespace Microsoft.AspNetCore.WebUtilities
                 throw new ArgumentOutOfRangeException(nameof(count));
             }
 
-            if (_disposed)
-            {
-                throw new ObjectDisposedException(nameof(HttpRequestStreamReader));
-            }
-
-            var charsRead = 0;
-            while (count > 0)
-            {
-                var charsRemaining = _charsRead - _charBufferIndex;
-                if (charsRemaining == 0)
-                {
-                    charsRemaining = ReadIntoBuffer();
-                }
-
-                if (charsRemaining == 0)
-                {
-                    break;  // We're at EOF
-                }
-
-                if (charsRemaining > count)
-                {
-                    charsRemaining = count;
-                }
-
-                Buffer.BlockCopy(
-                    _charBuffer,
-                    _charBufferIndex * 2,
-                    buffer,
-                    (index + charsRead) * 2,
-                    charsRemaining * 2);
-                _charBufferIndex += charsRemaining;
-
-                charsRead += charsRemaining;
-                count -= charsRemaining;
-
-                // If we got back fewer chars than we asked for, then it's likely the underlying stream is blocked.
-                // Send the data back to the caller so they can process it.
-                if (_isBlocked)
-                {
-                    break;
-                }
-            }
-
-            return charsRead;
+            var span = new Span<char>(buffer, index, count);
+            return Read(span);
         }
 
         public override int Read(Span<char> buffer)
@@ -206,6 +164,11 @@ namespace Microsoft.AspNetCore.WebUtilities
             if (buffer == null)
             {
                 throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(HttpRequestStreamReader));
             }
 
             var count = buffer.Length;

--- a/src/Http/WebUtilities/src/HttpRequestStreamReader.cs
+++ b/src/Http/WebUtilities/src/HttpRequestStreamReader.cs
@@ -6,6 +6,7 @@ using System.Buffers;
 using System.Diagnostics;
 using System.IO;
 using System.Text;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace Microsoft.AspNetCore.WebUtilities
@@ -251,7 +252,7 @@ namespace Microsoft.AspNetCore.WebUtilities
             return charsRead;
         }
 
-        public override async Task<int> ReadAsync(char[] buffer, int index, int count)
+        public override Task<int> ReadAsync(char[] buffer, int index, int count)
         {
             if (buffer == null)
             {
@@ -268,6 +269,12 @@ namespace Microsoft.AspNetCore.WebUtilities
                 throw new ArgumentOutOfRangeException(nameof(count));
             }
 
+            var memory = new Memory<char>(buffer, index, count);
+            return ReadAsync(memory).AsTask();
+        }
+
+        public override async ValueTask<int> ReadAsync(Memory<char> buffer, CancellationToken cancellationToken = default)
+        {
             if (_disposed)
             {
                 throw new ObjectDisposedException(nameof(HttpRequestStreamReader));
@@ -278,14 +285,16 @@ namespace Microsoft.AspNetCore.WebUtilities
                 return 0;
             }
 
+            var count = buffer.Length;
+
             var charsRead = 0;
             while (count > 0)
             {
                 // n is the characters available in _charBuffer
-                var n = _charsRead - _charBufferIndex;
+                var charsAvailable = _charsRead - _charBufferIndex;
 
                 // charBuffer is empty, let's read from the stream
-                if (n == 0)
+                if (charsAvailable == 0)
                 {
                     _charsRead = 0;
                     _charBufferIndex = 0;
@@ -295,7 +304,7 @@ namespace Microsoft.AspNetCore.WebUtilities
                     // We break out of the loop if the stream is blocked (EOF is reached).
                     do
                     {
-                        Debug.Assert(n == 0);
+                        Debug.Assert(charsAvailable == 0);
                         _bytesRead = await _stream.ReadAsync(
                             _byteBuffer,
                             0,
@@ -309,45 +318,47 @@ namespace Microsoft.AspNetCore.WebUtilities
                         // _isBlocked == whether we read fewer bytes than we asked for.
                         _isBlocked = (_bytesRead < _byteBufferSize);
 
-                        Debug.Assert(n == 0);
+                        Debug.Assert(charsAvailable == 0);
 
                         _charBufferIndex = 0;
-                        n = _decoder.GetChars(
+                        charsAvailable = _decoder.GetChars(
                             _byteBuffer,
                             0,
                             _bytesRead,
                             _charBuffer,
                             0);
 
-                        Debug.Assert(n > 0);
+                        Debug.Assert(charsAvailable > 0);
 
-                        _charsRead += n; // Number of chars in StreamReader's buffer.
+                        _charsRead += charsAvailable; // Number of chars in StreamReader's buffer.
                     }
-                    while (n == 0);
+                    while (charsAvailable == 0);
 
-                    if (n == 0)
+                    if (charsAvailable == 0)
                     {
                         break; // We're at EOF
                     }
                 }
 
                 // Got more chars in charBuffer than the user requested
-                if (n > count)
+                if (charsAvailable > count)
                 {
-                    n = count;
+                    charsAvailable = count;
                 }
 
-                Buffer.BlockCopy(
-                    _charBuffer,
-                    _charBufferIndex * 2,
-                    buffer,
-                    (index + charsRead) * 2,
-                    n * 2);
+                var source = new Memory<char>(_charBuffer, _charBufferIndex, charsAvailable);
+                source.CopyTo(buffer);
 
-                _charBufferIndex += n;
+                if (charsAvailable < count)
+                {
+                    // update the buffer to the remaining portion
+                    buffer = buffer.Slice(charsAvailable);
+                }
 
-                charsRead += n;
-                count -= n;
+                _charBufferIndex += charsAvailable;
+
+                charsRead += charsAvailable;
+                count -= charsAvailable;
 
                 // This function shouldn't block for an indefinite amount of time,
                 // or reading from a network stream won't work right.  If we got

--- a/src/Http/WebUtilities/src/HttpRequestStreamReader.cs
+++ b/src/Http/WebUtilities/src/HttpRequestStreamReader.cs
@@ -200,6 +200,57 @@ namespace Microsoft.AspNetCore.WebUtilities
             return charsRead;
         }
 
+        public override int Read(Span<char> buffer)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            var count = buffer.Length;
+            var charsRead = 0;
+            while (count > 0)
+            {
+                var charsRemaining = _charsRead - _charBufferIndex;
+                if (charsRemaining == 0)
+                {
+                    charsRemaining = ReadIntoBuffer();
+                }
+
+                if (charsRemaining == 0)
+                {
+                    break;  // We're at EOF
+                }
+
+                if (charsRemaining > count)
+                {
+                    charsRemaining = count;
+                }
+
+                var source = new ReadOnlySpan<char>(_charBuffer, _charBufferIndex, charsRemaining);
+                source.CopyTo(buffer);
+
+                _charBufferIndex += charsRemaining;
+
+                charsRead += charsRemaining;
+                count -= charsRemaining;
+
+                if (count > 0)
+                {
+                    buffer = buffer.Slice(charsRemaining, count);
+                }
+
+                // If we got back fewer chars than we asked for, then it's likely the underlying stream is blocked.
+                // Send the data back to the caller so they can process it.
+                if (_isBlocked)
+                {
+                    break;
+                }
+            }
+
+            return charsRead;
+        }
+
         public override async Task<int> ReadAsync(char[] buffer, int index, int count)
         {
             if (buffer == null)

--- a/src/Http/WebUtilities/test/HttpRequestStreamReaderTest.cs
+++ b/src/Http/WebUtilities/test/HttpRequestStreamReaderTest.cs
@@ -13,7 +13,7 @@ using Xunit;
 
 namespace Microsoft.AspNetCore.WebUtilities
 {
-    public class HttpResponseStreamReaderTest
+    public class HttpRequestStreamReaderTest
     {
         private static readonly char[] CharData = new char[]
         {
@@ -118,7 +118,7 @@ namespace Microsoft.AspNetCore.WebUtilities
         }
 
         [Fact]
-        public static async Task Read_ReadInTwoChunks()
+        public static async Task ReadAsync_ReadInTwoChunks()
         {
             // Arrange
             var reader = CreateReader();
@@ -224,6 +224,44 @@ namespace Microsoft.AspNetCore.WebUtilities
 
             // Act
             var read = reader.Read(span);
+
+            // Assert
+            Assert.Equal(chars.Length, read);
+            for (var i = 0; i < CharData.Length; i++)
+            {
+                Assert.Equal(CharData[i], chars[i]);
+            }
+        }
+
+        [Fact]
+        public async static Task ReadAsync_Memory_ReadAllCharactersAtOnce()
+        {
+            // Arrange
+            var reader = CreateReader();
+            var chars = new char[CharData.Length];
+            var memory = new Memory<char>(chars);
+
+            // Act
+            var read = await reader.ReadAsync(memory);
+
+            // Assert
+            Assert.Equal(chars.Length, read);
+            for (var i = 0; i < CharData.Length; i++)
+            {
+                Assert.Equal(CharData[i], chars[i]);
+            }
+        }
+
+        [Fact]
+        public async static Task ReadAsync_Memory_WithMoreDataThanInternalBufferSize()
+        {
+            // Arrange
+            var reader = CreateReader(10);
+            var chars = new char[CharData.Length];
+            var memory = new Memory<char>(chars);
+
+            // Act
+            var read = await reader.ReadAsync(memory);
 
             // Assert
             Assert.Equal(chars.Length, read);

--- a/src/Http/WebUtilities/test/HttpRequestStreamReaderTest.cs
+++ b/src/Http/WebUtilities/test/HttpRequestStreamReaderTest.cs
@@ -195,6 +195,44 @@ namespace Microsoft.AspNetCore.WebUtilities
             Assert.Null(eol);
         }
 
+        [Fact]
+        public static void Read_Span_ReadAllCharactersAtOnce()
+        {
+            // Arrange
+            var reader = CreateReader();
+            var chars = new char[CharData.Length];
+            var span = new Span<char>(chars);
+
+            // Act
+            var read = reader.Read(span);
+
+            // Assert
+            Assert.Equal(chars.Length, read);
+            for (var i = 0; i < CharData.Length; i++)
+            {
+                Assert.Equal(CharData[i], chars[i]);
+            }
+        }
+
+        [Fact]
+        public static void Read_Span_WithMoreDataThanInternalBufferSize()
+        {
+            // Arrange
+            var reader = CreateReader(10);
+            var chars = new char[CharData.Length];
+            var span = new Span<char>(chars);
+
+            // Act
+            var read = reader.Read(span);
+
+            // Assert
+            Assert.Equal(chars.Length, read);
+            for (var i = 0; i < CharData.Length; i++)
+            {
+                Assert.Equal(CharData[i], chars[i]);
+            }
+        }
+
         [Theory]
         [MemberData(nameof(HttpRequestNullData))]
         public static void NullInputsInConstructor_ExpectArgumentNullException(Stream stream, Encoding encoding, ArrayPool<byte> bytePool, ArrayPool<char> charPool)
@@ -253,15 +291,27 @@ namespace Microsoft.AspNetCore.WebUtilities
                 return httpRequestStreamReader.ReadAsync(new char[10], 0, 1);
             });
         }
+
         private static HttpRequestStreamReader CreateReader()
+        {
+            MemoryStream stream = CreateStream();
+            return new HttpRequestStreamReader(stream, Encoding.UTF8);
+        }
+
+        private static HttpRequestStreamReader CreateReader(int bufferSize)
+        {
+            MemoryStream stream = CreateStream();
+            return new HttpRequestStreamReader(stream, Encoding.UTF8, bufferSize);
+        }
+
+        private static MemoryStream CreateStream()
         {
             var stream = new MemoryStream();
             var writer = new StreamWriter(stream);
             writer.Write(CharData);
             writer.Flush();
             stream.Position = 0;
-
-            return new HttpRequestStreamReader(stream, Encoding.UTF8);
+            return stream;
         }
 
         private static MemoryStream GetSmallStream()

--- a/src/Http/WebUtilities/test/HttpRequestStreamReaderTest.cs
+++ b/src/Http/WebUtilities/test/HttpRequestStreamReaderTest.cs
@@ -318,16 +318,14 @@ namespace Microsoft.AspNetCore.WebUtilities
             });
         }
 
-        [Fact]
-        public static async Task StreamDisposed_ExpectObjectDisposedExceptionAsync()
+        [Theory]
+        [MemberData(nameof(HttpRequestDisposeDataAsync))]
+        public static async Task StreamDisposed_ExpectObjectDisposedExceptionAsync(Func<HttpRequestStreamReader, Task> action)
         {
             var httpRequestStreamReader = new HttpRequestStreamReader(new MemoryStream(), Encoding.UTF8, 10, ArrayPool<byte>.Shared, ArrayPool<char>.Shared);
             httpRequestStreamReader.Dispose();
 
-            await Assert.ThrowsAsync<ObjectDisposedException>(() =>
-            {
-                return httpRequestStreamReader.ReadAsync(new char[10], 0, 1);
-            });
+            await Assert.ThrowsAsync<ObjectDisposedException>(() => action(httpRequestStreamReader));
         }
 
         private static HttpRequestStreamReader CreateReader()
@@ -390,12 +388,28 @@ namespace Microsoft.AspNetCore.WebUtilities
             {
                  var res = httpRequestStreamReader.Read(new char[10], 0, 1);
             })};
+            yield return new object[] { new Action<HttpRequestStreamReader>((httpRequestStreamReader) =>
+            {
+                 var res = httpRequestStreamReader.Read(new Span<char>(new char[10], 0, 1));
+            })};
 
             yield return new object[] { new Action<HttpRequestStreamReader>((httpRequestStreamReader) =>
             {
                 var res = httpRequestStreamReader.Peek();
             })};
 
+        }
+
+        public static IEnumerable<object[]> HttpRequestDisposeDataAsync()
+        {
+            yield return new object[] { new Func<HttpRequestStreamReader, Task>(async (httpRequestStreamReader) =>
+            {
+                 await httpRequestStreamReader.ReadAsync(new char[10], 0, 1);
+            })};
+            yield return new object[] { new Func<HttpRequestStreamReader, Task>(async (httpRequestStreamReader) =>
+            {
+                 await httpRequestStreamReader.ReadAsync(new Memory<char>(new char[10], 0, 1));
+            })};
         }
     }
 }

